File: 2337-move-pieces-to-obtain-a-string/2337-move-pieces-to-obtain-a-string.js
var canChange = function (start, target) {
    let x = 0, y = 0;

    while (x < start.length || y < target.length) {
        while (x < start.length && start[x] === '_') {
            x++;
        }
        while (y < target.length && target[y] === '_') {
            y++;
        }
        if (start[x] !== target[y]) return false;

        if (start[x] === 'L' && x < y) return false;

        if (start[x] === 'R' && y < x) return false;

        x++;
        y++;
    }
    return true;
};

File: 2337-move-pieces-to-obtain-a-string/2337-move-pieces-to-obtain-a-string.cpp
class Solution {
public:
    bool canChange(string start, string target) {
        int n = start.length();
        int l=0, r=0;
        for(int i=0; i<n; i++){
            if(start[i] == 'L'){
                l--;
            }
            else if(start[i] == 'R'){
                r++;
                if(l!=0) return false;
            }
            if(target[i] == 'L'){
                l++;
                if(r!=0) return false;
            }
            else if(target[i] == 'R'){
                r--;
            }

            if(l<0 || r<0) return false;
        }
        return l==0 && r==0;
    }
};
